### DISCUSSION TITLE
How to do Key Sequences?

### DISCUSSION THREAD

**Discussion by kaygdotorg** [evan-liu/karabiner.ts#123] [Q&A] [answered]
I am really inspired by https://github.com/mxstbr/karabiner, the way he uses hyper sub layers. I used to do something similar with Doom Emacs but it's not exactly the same. Instead of pressing Hyper + R (raycast) + S (let's say spotify), I would like to create: Hyper + R (raycast) > (and then) S (spotify), as it's very incovenient for me to press 3 key shortcuts. 

However I am also too dumb to understand typescript very well (I've never written javascript or typescript in my life) so I was wondering if you could provide a few examples on how to do it. Thank you!

**Comment by evan-liu** [comment]
Hi @kayg04, sorry I do not know how to do that with Karabiner-Element. Not sure if those could be helpful: 

- [hyperLayer()](https://evan-liu.github.io/karabiner.ts/rules/hyper-layer)
- [duoLayer()](https://evan-liu.github.io/karabiner.ts/rules/duo-layer)

**Reply by kaygdotorg** [reply, marked as answer]
Sadly, none of the two work for me. :/ 

Thank you, your tool does make maintaining a karabiner config much easier.  
I guess I'll keep using BetterTouchTool for key sequences.

**Comment by evan-liu** [comment]
Hi @kayg04, I added a new feature ["leader mode"](https://evan-liu.github.io/karabiner.ts/rules/leader-mode), which could work for your needs. Please let me know if you'd like to give it a try. 

```typescript
hyperLayer('r') // Hyper + r
  .description('Raycast')
  .leaderMode()
  .notification() // Notification is not required but highly recommended
  .manipulators({
    s: toApp('Spotify'),
    // ... other apps
  }),
```

**Reply by kaygdotorg** [reply]
Hey, I am coming back to this now. Let me try it out and get back to you. Thanks a lot for implementing this.

**Reply by kaygdotorg** [reply]
Hi Evan,

This does not seem to be working for me. Press Caps-Lock + r -> s does nothing, I also added a notification to confirm but nope, nothing:

```
  hyperLayer('r') // Hyper + r
  .description('Raycast')
  .leaderMode({ escape: ['spacebar', 'return_or_enter'] })
  .notification("ke test") // Notification is not required but highly recommended
  .manipulators({
    s: toApp('Spotify'),
  }),
```

Is there anything I could be missing?

**Reply by kaygdotorg** [reply]
Alright, it was my dumb brain, I didn't change the profile name. This works! 

I'll report back with some more testing. Thank you again!

**Reply by kaygdotorg** [reply]
So how would I open a url here? I don't see anything like an open function?

**Reply by evan-liu** [reply]
to$(`open "https://some.url"`)
